tracks/python/exercises/book-store/mentoring.md

```diff
@@ -0,0 +1,301 @@
+    )
+
+def combo_price(combo):
+    num_books = len(combo)
+    num_titles = len(set(combo))
+    assert num_books == num_titles, "combo titles must be unique"
+    price_per_book = FULL_PRICE * (1 - DISCOUNTS[len(combo)])
+    return num_books * price_per_book
+
+def remove(quantities, items):
+    return tuple(
+        quantity - items.count(title)
+        for (title, quantity) in zip(TITLES, quantities)
+    )
+```
+
+
+Why break it up like this?
+
+-   Extensibility - e.g. it would be easy to add different prices by changing the implementation
+    of `combo_price`;
+
+-   Code reuse - the same helper functions are used in the iterative solutions below;
+
+-   Testing - we could write unit tests for each of the helper functions;
+
+-   Readability - the encapsulation provided by the helper functions makes the main function more
+    declarative;
+
+-   Functional programming - less mutability make the code less taxing to read (i.e. it doesn't
+    use as much of your working memory as you don't need to mentally simulate the evolution of the
+    state of the program).
+
+
+What about performance?
+
+-   Currently, this implementation is extremely slow. But, here's an easy optimization: use a cache.
+
+-   Why does this produce such a significant improvement in performance? There are different ways to
+    end up at the same point in the search space; it turns tree-search into graph-search. This is
+    the essence of dynamic programming.
+
+-   The runtime of the uncached function grows exponentially w.r.t.  `sum(quantities)`; but the
+    cached function's cache certainly grows no larger than `product(quantities)`.
+
+
+You could implement the cache yourself:
+
+```python
+def best_price_for(quantities, _cache=dict()):
+    """Find the best price for the given quantities of titles."""
+    if quantities not in _cache:
+        _cache[quantities] = ...
+    return _cache[quantities]
+```
+
+(Is there something suprising about the behavior of `_cache=dict()`? Here, this behavior is
+exploited intentionally, but it can be source of bugs.)
+
+
+...or you could use the built-in caching decorator:
+
+```python
+from functools import lru_cache
+
+@lru_cache(None)
+def best_price_for(quantities):
+    """Find the best price for the given quantities of titles."""
+    ...
+```
+
+The latter is preferable for a couple of reasons: it doesn't re-invent the wheel, and it happens to
+also be thread-safe (perhaps a topic of discussion for the more advanced mentee).
+
+
+Here's another easy optimization:
+
+-   To improve any d.p. algorithm, try to reduce the size of the search space.
+
+-   One way to do this is to find classes of states that produce the same result, and represent each
+    class with a canonical form.
+
+-   Since prices are all the same, how much we have of which title doesn't matter, so:
+        best_price_for(quantities) == best_price_for(tuple(sorted(quantities)))
+
+
+Put into practice:
+
+```python
+@lru_cache(None)
```

Comment:
If only to get the reader to think about what's happening use the keyword form, ie `@lru_cache(max_size=None)`.